contracts/epochs/FlowDKG.cdc

```diff
@@ -0,0 +1,274 @@
+        /// Resets all the fields for tracking the current DKG process
+        /// and sets the given node IDs as registered
+        pub fun startDKG(nodeIDs: [String]) {
+            FlowDKG.dkgEnabled = true
+
+            FlowDKG.nodeFinalSubmission = {}
+            for id in nodeIDs {
+                FlowDKG.nodeFinalSubmission[id] = []
+            }
+
+            FlowDKG.whiteboardMessages = []
+
+            FlowDKG.uniqueFinalSubmissions = []
+
+            FlowDKG.uniqueFinalSubmissionCount = {}
+
+            emit StartDKG()
+        }
+
+        pub fun endDKG() {
+            FlowDKG.dkgEnabled = false
+
+            emit EndDKG(finalSubmission: FlowDKG.dkgCompleted())
+        }
+    }
+
+    access(account) fun submissionsEqual(_ existingSubmission: [String], _ submission: [String]): Bool {
+
+        // If the submission length is different than the one being compared to, it is not equal
+        if submission.length != existingSubmission.length {
+            return false
+        }
+
+        var index = 0
+
+        // Check each key in the submiission to make sure that it matches
+        // the existing one
+        for key in submission {
+            // If a key length is incorrect, it is an invalid submission
+            if key.length != FlowDKG.submissionKeyLength {
+                panic("Submission key length is not correct!")
+            }
+
+            // if a key is different, stop checking this submission
+            // and move on to the next one
+            if key != existingSubmission[index] {
+                return false
+            }
+
+            index = index + 1
+        }
+
+        return true
+
+    }
+
+    pub fun participantIsRegistered(_ nodeID: String): Bool {
+        return FlowDKG.nodeFinalSubmission[nodeID] != nil
+    }
+
+    pub fun participantIsClaimed(_ nodeID: String): Bool? {
+        return FlowDKG.nodeClaimed[nodeID]
+    }
+
+    // specify last X messages in the transaction
+    pub fun getWhiteBoardMessages(): [Message] {
+        return self.whiteboardMessages
+    }
+
+    /// Returns whether this node has successfully submitted a final submission for this epoch.
+    /// Returns nil if the node is not a participant
+    /// Returns empty array if the node is a participant or hasn't submitted
+    /// Returns a submission array if the node has submitted
+    pub fun nodeHasSubmitted(_ nodeID: String): [String]? {
+        return self.nodeFinalSubmission[nodeID]
+    }
+
+    /// Get the list of all the consensus node IDs participating
+    pub fun getConsensusNodeIDs(): [String] {
+        return self.nodeFinalSubmission.keys
+    }
+
+    /// Get the array of all the unique final submissions
+    pub fun getFinalSubmissions(): [[String]] {
+        return self.uniqueFinalSubmissions
+    }
+
+    /// Returns the final set of keys if any one set of keys has more than 50% submissions
+    /// Returns nil if not found (incomplete)
+    pub fun dkgCompleted(): [String]? {
```

Comment:
I'm wondering when this function is called. Is there a deadline for nodes to submit their submissions after which `dkgCompleted` is called?

Reply:
I was intending that the main epoch contract would poll this function every block and when it and the qc is complete, it would be able to end the epoch setup phase. We know it is complete when this returns an array instead of `nil`. Does that make sense?

Reply:
Yes that makes sense 👌